pkg/csi_driver/multishare_ops_manager.go

```diff
@@ -651,16 +648,83 @@ func containsOpWithInstanceTargetPrefix(instance *file.MultishareInstance, ops [
 	return nil, nil
 }
 
-func (m *MultishareOpsManager) listInstanceForStorageClassPrefix(ctx context.Context, prefix string) ([]*file.MultishareInstance, error) {
+// listMatchedInstances will list all instances in current project,
+// but only matched instances will be returned.
+func (m *MultishareOpsManager) listMatchedInstances(ctx context.Context, req *csi.CreateVolumeRequest, target *file.MultishareInstance) ([]*file.MultishareInstance, error) {
 	instances, err := m.cloud.File.ListMultishareInstances(ctx, &file.ListFilter{Project: m.cloud.Project, Location: "-"})
 	if err != nil {
 		return nil, err
 	}
 	var finalInstances []*file.MultishareInstance
 	for _, i := range instances {
-		if val, ok := i.Labels[util.ParamMultishareInstanceScLabelKey]; ok && val == prefix {
+		matched, err := isMatchedInstance(i, target, req)
+		if err != nil {
+			return nil, err
+		}
+		if matched {
 			finalInstances = append(finalInstances, i)
 		}
 	}
 	return finalInstances, nil
 }
+
+// isIpWithinRange checks if an ip address is within the given ip range.
+func isIpWithinRange(ipAddress, ipRange string) (bool, error) {
+	_, ipnet, err := net.ParseCIDR(ipRange)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse cidr range %s: %v", ipRange, err)
+	}
+	return ipnet.Contains(net.ParseIP(ipAddress)), nil
+}
+
+// A source instance will be considered as "matched" with the target instance
+// if and only if the following requirements were met:
+// 1. Both source and target instance should have a label with key
+//    "storage_gke_io_storage-class-id", and the value should be the same.
+// 2. (Check if exists) The ip address of the target instance should be
+//    within the ip range specified in "reserved-ipv4-cidr".
+// 3. (Check if exists) The ip address of the target instance should be
+//    within the ip range specified in "reserved-ip-range".
+// 4. Both source and target instance should be in the same location.
+// 5. Both source and target instance should be under the same tier.
+// 6. Both source and target instance should be in the same VPC network.
+// 7, Both source and target instance should have the same connect mode.
+// 8. Both source and target instance should have the same KmsKeyName.
+func isMatchedInstance(source, target *file.MultishareInstance, req *csi.CreateVolumeRequest) (bool, error) {
```

Comment:
why do you need both `target` and `req` here? can't you read everything you need from `req`

Reply:
Good question. Consider the following workflow in CreateVolume:
1. Generate a dummy instance based on csi req. We verify and parse all params from csi req except for ip reservations [here](https://github.com/kubernetes-sigs/gcp-filestore-csi-driver/blob/1c2176ac2279e77ab42540b20470b1db7775213b/pkg/csi_driver/multishare_controller.go#L356). The generated instance will be used as `target`. 
2. ListMatchedInstances are called to get instances that are matched with the `target` from step#1.
3. If none eligible instance was found, we parse the CIDR from csi req, stamp it to `target` and file an instance creation op.
Because at step#1, no ip information are stamped to `target` instance, we need to extract the ip info from csi req. While `target` is still required because we don't want to verify and parse the params again from csi req, since it has already been done at step#1.